homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,273 @@
+
+            except (requests.RequestException, IOError):
+                camera.is_streaming = False
+                camera.update_ha_state()
+
+        else:
+            handler.send_response(HTTP_NOT_FOUND)
+
+        camera.is_streaming = False
+
+    hass.http.register_path(
+        'GET',
+        re.compile(
+            r'/api/camera_proxy_stream/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _proxy_camera_mjpeg_stream,
+        require_auth=True)
+
+
+class Camera(Entity):
+    """ The base class for camera components """
+
+    @property
+    # pylint: disable=no-self-use
+    def is_recording(self):
+        """ Returns true if the device is recording """
+        return False
+
+    @property
+    # pylint: disable=no-self-use
+    def is_streaming(self):
+        """ Returns true if the device is streaming """
+        return False
+
+    @is_streaming.setter
+    def is_streaming(self, value):
+        """ Set this to true when streaming begins """
+        pass
+
+    @property
+    # pylint: disable=no-self-use
+    def brand(self):
+        """ Should return a string of the camera brand """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def model(self):
+        """ Returns string of camera model """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def base_url(self):
+        """ Return the configured base URL for the camera """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def image_url(self):
+        """ Return the still image segment of the URL """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def device_info(self):
+        """ Get the configuration object """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def username(self):
+        """ Get the configured username """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def password(self):
+        """ Get the configured password """
+        return None
+
+    @property
+    # pylint: disable=no-self-use
+    def still_image_url(self):
+        """ Get the URL of a camera still image """
+        return None
+
+    def get_camera_image(self):
+        """ Return bytes of camera image """
+        raise NotImplementedError()
+
+    @property
+    def state(self):
+        """ Returns the state of the entity. """
+        if self.is_recording:
+            return STATE_RECORDING
+        elif self.is_streaming:
+            return STATE_STREAMING
+        else:
+            return STATE_IDLE
+
+    @property
+    def state_attributes(self):
+        """ Returns optional state attributes. """
+        attr = super().state_attributes
+        attr['model_name'] = self.device_info.get('model', 'generic')
+        attr['brand'] = self.device_info.get('brand', 'generic')
+        attr['still_image_url'] = '/api/camera_proxy/' + self.entity_id
```

Comment:
Could you extract the attribute keys as constants and put them on top of the file. Also, please don't use string concatenation but string formatting:
```python
attr[ATTR_STILL_IMAGE_URL] = CAMERA_PROXY_URL.format(self.entity_id)
```